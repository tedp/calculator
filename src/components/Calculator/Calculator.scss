@import '../../styles/styles.scss';

$module: "Calculator";

.#{$module} {
  padding: $padding * 2;
  @include box-shadow();
  border-radius: 1px;
  width: $width;
  margin: auto;
}

