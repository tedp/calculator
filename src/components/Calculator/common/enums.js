const KeyType = Object.freeze({
  numeric: 'numeric',
  dot: 'dot',
  operator: 'operator',
  clear: 'clear',
  result: 'result',
});

const DisplayState = Object.freeze({
  numeric: 'numeric',
  decimal: 'decimal',
  dot: 'dot',
  operator: 'operator',
});

const CalcState = Object.freeze({
  resulted: 'resulted',
  cleared: 'cleared',
  input: 'input',
});

export { KeyType, CalcState, DisplayState };
