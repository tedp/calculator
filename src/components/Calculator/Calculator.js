import React from 'react';
import './Calculator.scss';
import { CalcState, KeyType, DisplayState } from './common/enums';
import ResultDisplay from './ResultDisplay/ResultDisplay';
import CalculatorInput from './CalculatorInput/CalculatorInput';
import calculateState from './functions';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.handleInput = this.handleInput.bind(this);

    this.state = {
      result: { display: '0', keyTypes: [KeyType.numeric], displayStates: [DisplayState.numeric] },
      calcState: CalcState.cleared,
      displayState: DisplayState.numeric,
      lastKeyType: KeyType.numeric,
    };
  }

  handleInput(button) {
    const state = calculateState(this.state, button.keyValue, button.keyType);
    this.setState(state);
  }

  render() {
    return (
      <div className="Calculator">
        <ResultDisplay displayValue={this.state.result.display} />
        <CalculatorInput onInput={this.handleInput} />
      </div>
    );
  }
}

export default Calculator;
