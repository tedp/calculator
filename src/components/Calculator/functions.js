import { CalcState, KeyType, DisplayState } from './common/enums';

function calculateState(state, keyValue, keyType) {
  function calculateDisplayState(currentDisplayState, keyType) {
    const nextCombinations = {
      [DisplayState.numeric + KeyType.numeric]: DisplayState.numeric,
      [DisplayState.numeric + KeyType.operator]: DisplayState.operator,
      [DisplayState.numeric + KeyType.dot]: DisplayState.dot,

      [DisplayState.dot + KeyType.numeric]: DisplayState.decimal,

      [DisplayState.decimal + KeyType.numeric]: DisplayState.decimal,
      [DisplayState.decimal + KeyType.operator]: DisplayState.operator,

      [DisplayState.operator + KeyType.numeric]: DisplayState.numeric,
    };

    const combination = currentDisplayState + keyType;

    return nextCombinations[combination];
  }

  function initialState(state, keyValue) {
    return {
      ...state,
      result: {
        display: keyValue,
        keyTypes: [KeyType.numeric],
        displayStates: [DisplayState.numeric],
      },
      calcState: CalcState.input,
      displayState: DisplayState.numeric,
    };
  }

  function concat(state, keyValue, keyType) {
    const display = (state.result.display += keyValue);
    const keyTypes = [...state.result.keyTypes, keyType];
    const displayState = calculateDisplayState(state.displayState, keyType);
    const displayStates = [...state.result.displayStates, displayState];
    return {
      ...state,
      displayState,
      result: { display, keyTypes, displayStates },
      calcState: CalcState.input,
    };
  }

  function clearLast(state) {
    let display;
    let keyTypes;
    let displayStates;
    let calcState;
    let displayState;

    if (state.result.display.length > 1) {
      display = state.result.display.slice(0, -1);
      keyTypes = state.result.keyTypes.slice(0, -1);
      calcState = state.calcState;
      displayStates = state.result.displayStates.slice(0, -1);
      displayState = displayStates[displayStates.length - 1];
    } else {
      display = '0';
      keyTypes = [KeyType.numeric];
      calcState = CalcState.cleared;
      displayState = DisplayState.numeric;
      displayStates = [displayState];
    }

    return {
      ...state,
      result: { display, keyTypes, displayStates },
      calcState,
      displayState,
    };
  }

  function calulateResultState(state) {
    function calculateResult(input) {
      function calculate(operator, operators, numbers, calcFunction) {
        let workingOperator = operators.indexOf(operator);
        while (workingOperator !== -1) {
          const result = calcFunction(numbers[workingOperator], numbers[workingOperator + 1]);
          numbers.splice(workingOperator, 2, result);
          operators.splice(workingOperator, 1);
          workingOperator = operators.indexOf(operator);
        }
      }

      const inputString = String(input);
      const numbers = inputString.split(/\+|-|×|÷/g);
      const operators = inputString.replace(/[0-9]|\./g, '').split('');

      // Calculate the result using the operators in the correct order
      calculate('÷', operators, numbers, (leftNumber, rightNumber) => leftNumber / rightNumber);
      calculate('×', operators, numbers, (leftNumber, rightNumber) => leftNumber * rightNumber);
      calculate('-', operators, numbers, (leftNumber, rightNumber) => leftNumber - rightNumber);
      // Use of parseFloat to avoid string concatenation
      calculate('+', operators, numbers, (leftNumber, rightNumber) => parseFloat(leftNumber) + parseFloat(rightNumber));

      return numbers[0];
    }

    const display = calculateResult(state.result.display);
    const keyTypes = [KeyType.numeric];
    const displayStates = [DisplayState.numeric];
    
    return {
      ...state,
      result: { display, keyTypes, displayStates },
      calcState: CalcState.resulted,
      displayState: DisplayState.numeric,
    };
  }

  const validCombinations = [
    DisplayState.numeric + KeyType.numeric,
    DisplayState.numeric + KeyType.operator,
    DisplayState.numeric + KeyType.dot,

    DisplayState.dot + KeyType.numeric,

    DisplayState.decimal + KeyType.numeric,
    DisplayState.decimal + KeyType.operator,

    DisplayState.operator + KeyType.numeric,
  ];

  if (keyType === KeyType.result) {
    return calulateResultState(state);
  }
  if (keyType === KeyType.clear) {
    return clearLast(state);
  }
  if (
    (state.calcState === CalcState.cleared
      || state.calcState === CalcState.resulted)
    && keyType === KeyType.numeric
  ) {
    return initialState(state, keyValue, keyType);
  }

  const combination = state.displayState + keyType;

  if (validCombinations.includes(combination)) {
    return concat(state, keyValue, keyType);
  }

  return state;
}

export default calculateState;
