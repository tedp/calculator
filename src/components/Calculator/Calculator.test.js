import React from 'react';
import { shallow, mount } from 'enzyme';
import Calculator from './Calculator';
import { KeyType } from './common/enums';

describe('CalculatorInput', () => {
  it('should render correctly', () => {
    const component = shallow(<Calculator />);

    expect(component).toMatchSnapshot();
  });

  it('should show a value of `9` in the display', () => {
    const component = mount(<Calculator />);

    component
    .find('.centre-panel .Button')
    .first()
    .simulate('click');

    expect(component).toMatchSnapshot();

    component.unmount()
  });
});
