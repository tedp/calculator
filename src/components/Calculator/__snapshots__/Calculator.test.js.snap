// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorInput should render correctly 1`] = `
<div
  className="Calculator"
>
  <ResultDisplay
    displayValue="0"
  />
  <CalculatorInput
    onInput={[Function]}
  />
</div>
`;

exports[`CalculatorInput should show a value of \`9\` in the display 1`] = `
<Calculator>
  <div
    className="Calculator"
  >
    <ResultDisplay
      displayValue="9"
    >
      <div
        className="ResultDisplay"
      >
        9
      </div>
    </ResultDisplay>
    <CalculatorInput
      onInput={[Function]}
    >
      <div
        className="CalculatorInput"
      >
        <div
          className="top-panel"
        >
          <Button
            key="+"
            onClick={[Function]}
            value="+"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              +
            </div>
          </Button>
          <Button
            key="-"
            onClick={[Function]}
            value="-"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              -
            </div>
          </Button>
          <Button
            key="×"
            onClick={[Function]}
            value="×"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              ×
            </div>
          </Button>
          <Button
            key="÷"
            onClick={[Function]}
            value="÷"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              ÷
            </div>
          </Button>
        </div>
        <div
          className="centre-panel"
        >
          <Button
            key="9"
            onClick={[Function]}
            value="9"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              9
            </div>
          </Button>
          <Button
            key="8"
            onClick={[Function]}
            value="8"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              8
            </div>
          </Button>
          <Button
            key="7"
            onClick={[Function]}
            value="7"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              7
            </div>
          </Button>
          <Button
            key="6"
            onClick={[Function]}
            value="6"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              6
            </div>
          </Button>
          <Button
            key="5"
            onClick={[Function]}
            value="5"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              5
            </div>
          </Button>
          <Button
            key="4"
            onClick={[Function]}
            value="4"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              4
            </div>
          </Button>
          <Button
            key="3"
            onClick={[Function]}
            value="3"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              3
            </div>
          </Button>
          <Button
            key="2"
            onClick={[Function]}
            value="2"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              2
            </div>
          </Button>
          <Button
            key="1"
            onClick={[Function]}
            value="1"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              1
            </div>
          </Button>
          <Button
            key="0"
            onClick={[Function]}
            value="0"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              0
            </div>
          </Button>
          <Button
            key="."
            onClick={[Function]}
            value="."
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              .
            </div>
          </Button>
          <Button
            key="C"
            onClick={[Function]}
            value="C"
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              C
            </div>
          </Button>
        </div>
        <div
          className="bottom-panel"
          id="result"
        >
          <Button
            onClick={[Function]}
            value="="
          >
            <div
              className="Button"
              onClick={[Function]}
            >
              =
            </div>
          </Button>
        </div>
      </div>
    </CalculatorInput>
  </div>
</Calculator>
`;
